src/hamcrest/core/core/future.py

```diff
@@ -0,0 +1,138 @@
+else:
+    # Future is not a parametrised type in earlier version of python
+    FutureT = Union[asyncio.Future, Generator, Awaitable]
+
+
+class FutureException(BaseMatcher[asyncio.Future]):
+    def __init__(
+        self,
+        expected: Type[Exception],
+        pattern: Optional[str] = None,
+        matching: Optional[Matcher] = None,
+    ) -> None:
+        self.pattern = pattern
+        self.matcher = matching
+        self.expected = expected
+
+    def _matches(self, future: asyncio.Future) -> bool:
+        if not asyncio.isfuture(future):
+            return False
+
+        if not future.done():
+            return False
+
+        if future.cancelled():
+            return False
+
+        exc = future.exception()
+        if exc is None:
+            return False
+
+        if isinstance(exc, self.expected):
+            if self.pattern is not None:
+                if re.search(self.pattern, str(exc)) is None:
+                    return False
+            if self.matcher is not None:
+                if not self.matcher.matches(exc):
+                    return False
+            return True
+
+        return False
+
+    def describe_to(self, description: Description) -> None:
+        description.append_text("Expected a future with exception %s" % self.expected)
+
+    def describe_mismatch(self, future: asyncio.Future, description: Description) -> None:
+        if not asyncio.isfuture(future):
+            description.append_text("%s is not a future" % future)
+            return
+
+        if not future.done():
+            description.append_text("%s is not completed yet" % future)
+            return
+
+        if future.cancelled():
+            description.append_text("%s is cancelled" % future)
+            return
+
+        exc = future.exception()
+        if exc is None:
+            description.append_text("No exception raised.")
+        elif isinstance(exc, self.expected):
+            if self.pattern is not None or self.matcher is not None:
+                description.append_text("Correct assertion type raised, but ")
+                if self.pattern is not None:
+                    description.append_text('the expected pattern ("%s") ' % self.pattern)
+                if self.pattern is not None and self.matcher is not None:
+                    description.append_text("and ")
+                if self.matcher is not None:
+                    description.append_description_of(self.matcher)
+                    description.append_text(" ")
+                description.append_text('not found. Exception message was: "%s"' % str(exc))
+        else:
+            description.append_text("%r of type %s was raised instead" % (exc, type(exc)))
+
+    def describe_match(self, future: asyncio.Future, match_description: Description) -> None:
+        exc = future.exception()
+        match_description.append_text("%r of type %s was raised." % (exc, type(exc)))
+
+
+def future_exception(
```

Comment:
I wonder just from a literate coding PoV if this should be (or at least have as an alias) `asynchronously_raises` instead of `future_exception`)

Reply:
I'm a bit torn on this but I'll admit I mostly find the literal coding concerns to be a distraction.

Technically it can match a Future object no matter how you produce it and it's not entirely correct to say it *raises* the exception because it was already raised but captured.

I'm happy to defer this decision to you though :)

Reply:
what about `assert_that(actual_future, is(future_raising(ValueError)))`, for clarity I think I'd prefer to have future still in the name.

Also, oops. did not mean to forget about this pull request for 2 years. I'll rebase if there's some interest in this approach :)